docs/design/enclave-tls/design-zh.md

```diff
@@ -0,0 +1,359 @@
+
+总的来说，硬件可执行环境（HardWare Trusted Execution Environment， HW-TEE）中运行的应用都是可信的，包括TLS库，因此似乎不用对HW-TEE内的TLS库进行特别的考虑，但与HW-TEE通信的对端必须要验证工作负载是否运行在真实的HW-TEE中。常规的远程证明工程化实践只是对远程证明底层原语进行了封装，实际产出就是让通信双方能够安全可信地协商出共享secret，并将如何基于共享secret创建安全通信信道的工作留给了应用开发者。
+
+研究表明，RA和建立安全信道的逻辑必须要深度结合，否则易受中间人攻击的影响[1]。即使不将RA与TLS这么复杂的协议结合，至少也要能实现仅通过RA协议来协商出作为安全信道基础的共享密钥这一基本功能。但是仅做到共享秘密信息的程度是无法完全实现安全信道的。虽然基于共享秘密的方式能够实现安全信道，但是基于RA协议实现的安全信道的通信效率太低。针对这个问题，解决方法是将RA与能够提供安全信道的标准协议相结合，比如TLS协议。
+
+Thomas Knauth 提出将Intel SGX远程证明和TLS结合的方案[2]，但是该方案面临着以下挑战：
+- 如何灵活地支持不同的TLS库？
+- 如何适配不同的HW-TEE环境？
+- 如何让两个不同类型的HW-TEE互相验证对方的Evidence？
+
+为了解决以上挑战，我们提出了Enclave-TLS，一种支持异构硬件可执行环境的双向传输层安全性协议。Enclave-TLS在TLS的基础上增加了将TLS证书中的公钥与HW-TEE attestation quote绑定的能力，基于HW-TEE硬件为信任根，即可证明了对方是在可信平台上，又可以高效的传输数据。同时Enclave-TLS还具有以下优势：
+- 能够支持不同的HW-TEE类型。
+- 能够支持不同的TLS库。
+- 能够支持不同的密码学算法库。
+- 能够支持不同类型HW-TEE间的双向TLS认证。
+
+# 架构设计
+
+![architecture.png](architecture.png)
+
+Enclave-TLS的架构如上图所示，Enclave TLS 提供了五个API给上层的应用和服务使用，五个API能够实现可信安全信道的建立和数据的传输。Enclave TLS API的实现依赖于核心层与四类实例的实现。
+
+为了提供更好的安全性，核心层与四类实例插件都运行在HW TEE环境中。由于Enclave-TLS架构需要支持和灵活选择多种可能性（例如：TLS库，加密库，Enclave形态等），因此根据功能逻辑区分出不同的实例插件是必不可少的。下面简单介绍一下核心层和实例插件的作用。
+- **核心层**。负责整体的控制作用，控制数据流的流向。
+- **TLS Wrapper实例**。负责完成真正的TLS session管理和网络传输。
+- **Attester 实例**。从本地平台运行环境中收集证明材料，通常要有Enclave配合生成quote数据。
+- **Verifier 实例**。负责验证收到的各种格式的quote数据，可能收到来自另一个不同的机密计算硬件平台发来的quote。也可能存在verifier与attester必须是相同平台的情况，比如verifier使用SGX ECDSA QVE的情况：强制要求当前运行环境必须支持SGX ECDSA，否则无法启动QVE来验证SGX ECDSA quote。
+Crypto实例。负责配合其他实例完成与密码学算法有关的操作。
+
+Enclave-TLS的设计主要有以下优势：
+- 将RA与能够提供安全信道的标准协议相结合，以硬件作为信任根，扩展数据安全边界。
+- 支持异构HW-TEE节点的安全可信通信，不局限于平台。
+- 模块化设计，可扩展性强。
+
+## 以硬件作为信任根，扩展数据安全边界
+
+![x509.png](x509.png)
+
+现代的PKI为例，信任根是由 Aliyun、Google 和 Microsoft 等少数实体维护的根证书颁发机构列表，我们希望使用Hardware TEE作硬件为信任根（例如：Intel SGX，AMD SEV等），扩展数据安全的边界。为此，证明者需要在创建TLS连接阶段向验证者提供必要的证明材料，使得挑战者验证证明者是否正在可信的TEE（例如，Intel SGX）中运行。
+
+为了实现将远程证明集成到 TLS 握手中的目标，我们必须解决两个目标：
+- 需要将服务器的 Enclave-TLS 密钥绑定到 enclave 实例。
+- 服务器/证明者必须在 TLS 连接设置期间提供必要的证明证据，以说服客户端/挑战者它确实运行在真实的TEE平台中。
+
+具体做法如上图所示，我们以客户端连接服务端然后验证服务器的身份为场景。其中服务器是证明者，客户端是挑战者。具体流程如下：
+
+1. 服务端生成密钥对，并将公钥和enclave实例绑定。
+
+服务端启动之后，会创建一个enclave，在enclave中生成一个受保护的新的密钥对。Enclave-TLS将该公钥的hash值作为用户数据包含到Attestation报告中来绑定Enclave-TLS密钥和Enclave。该报告由证明者生成，并确保该报告确实是在真实的TEE平台上生成的。至此，Enclave-TLS公钥与enclave实例绑定。
+
+2. 服务端生成带有证明材料的x.509证书。
+
+我们引入了新的 X.509 扩展，每个扩展都有自己的唯一对象标识符 (OID)，从而将TEE平台的证明材料作为自定义 X.509 扩展嵌入服务器的证书中。 将证明材料添加到证书中比引入新的 TLS 扩展的侵入性要小，例如，引入新的 TLS 扩展需要对每个 TLS 实现进行修改。 通过扩展证书，我们不需要对现有 TLS 库进行任何更改，并且使用现成的hook机制来验证新的X.509 扩展。
+
+传统上扩展证书需要由 CA 对其进行签名。 但是，由于我们建议使用硬件TEE 作为信任根，我们可以简单地对证书进行自签名。 我们不依靠 CA 将域名绑定到服务器的身份/密钥，而是依靠硬件平台来提供身份。
+
+嵌入到 Enclave-TLS 证书中的证明材料根据Attester 实例的不同有所不同。具体信息需要参考不同的enclave实例的远程报告信息。
+
+3. 客户端验证证书
+
+Enclave-TLS 证书的接收者必须验证Evidence来确定特定平台上Enclave的可信度。 具体步骤因认证方法而异。 在协议方面，验证发生在标准 TLS 握手期间。 如果任何验证步骤失败，端点将中止握手以终止 TLS 连接。除此之外，验证者必须确保Enclave-TLS证书的公钥的hash确实存在于enclave的报告中，然后验证者将从证书中获取到的enclave的身份信息和验证者已有的基准信息进行比较。
+
+## 异构HW-TEE节点的安全可信通信
+
```

Comment:
If this function is not implemented or it is a future plan, please specify out it.